src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,411 @@
+            }
+        }
+
+        /*
+        Increment total hit count and validate if threshold is reached.
+         */
+        void incrementTotalHitCount() throws IOException {
+            totalHits++;
+            hitsThresholdChecker.incrementHitCount();
+            if (scoreMode().isExhaustive() == false
+                && getTotalHitsRelation() == TotalHits.Relation.EQUAL_TO
+                && hitsThresholdChecker.isThresholdReached()) {
+                // for the first time hitsThreshold is reached, notify all comparators about this
+                for (LeafFieldComparator comparator : comparators) {
+                    comparator.setHitsThresholdReached();
+                }
+                setTotalHitsRelation(TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO);
+            }
+        }
+
+        /*
+        Collect hit and add the value of the sort field in the comparator.
+         */
+        void collectHit(int doc, int hitsCollected, int subQueryNumber, float score) throws IOException {
+            // Startup transient: queue hasn't gathered numHits yet
+            int slot = hitsCollected - 1;
+            // Copy hit into queue
+            comparators[subQueryNumber].copy(slot, doc);
+            add(slot, doc, compoundScores[subQueryNumber], subQueryNumber, score);
+            if (queueFull[subQueryNumber]) {
+                comparators[subQueryNumber].setBottom(bottom.slot);
+            }
+        }
+
+        /*
+        // This hit is competitive - replace bottom element in queue & adjustTop
+         */
+        void collectCompetitiveHit(int doc, int subQueryNumber) throws IOException {
+            comparators[subQueryNumber].copy(bottom.slot, doc);
+            updateBottom(doc, compoundScores[subQueryNumber]);
+            comparators[subQueryNumber].setBottom(bottom.slot);
+        }
+
+        boolean thresholdCheck(int doc, int subQueryNumber) throws IOException {
+            if (collectedAllCompetitiveHits || reverseMul * comparators[subQueryNumber].compareBottom(doc) <= 0) {
+                // since docs are visited in doc Id order, if compare is 0, it means
+                // this document is larger than anything else in the queue, and
+                // therefore not competitive.
+                if (searchSortPartOfIndexSort) {
+                    if (hitsThresholdChecker.isThresholdReached()) {
+                        setTotalHitsRelation(TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO);
+                        throw new CollectionTerminatedException();
+                    } else {
+                        collectedAllCompetitiveHits = true;
+                    }
+                }
+                return true;
+            }
+            return false;
+        }
+
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        return hitsThresholdChecker.scoreMode();
+    }
+
+    /*
+      TopFieldDocs per subquery
+     */
+    protected TopFieldDocs topDocsPerQuery(
```

Comment:
any reason for making this protected? I see that it not getting used in base classes. plus do we recommend overriding this function in future by base classes if no, then we should make this as private.

Reply:
made this method private